rts/Sim/Objects/SolidObject.cpp

```diff
@@ -113,11 +113,12 @@ void CSolidObject::UpdatePhysicalState(float eps)
 	//   the height property is used for much fewer purposes
 	//   than radius, so less reliable for determining state
 	#define MASK_NOAIR (PSTATE_BIT_ONGROUND | PSTATE_BIT_INWATER | PSTATE_BIT_UNDERWATER | PSTATE_BIT_UNDERGROUND)
+	const float waterLevel = CGround::GetWaterLevel(pos.x, pos.z);
 	ps |= (PSTATE_BIT_ONGROUND    * ((   pos.y -         gh) <=  eps));
-	ps |= (PSTATE_BIT_INWATER     * ((   pos.y             ) <= 0.0f));
+	ps |= (PSTATE_BIT_INWATER     * ((   pos.y             ) <= waterLevel));
 //	ps |= (PSTATE_BIT_UNDERWATER  * ((   pos.y +     height) <  0.0f));
 //	ps |= (PSTATE_BIT_UNDERGROUND * ((   pos.y +     height) <    gh));
-	ps |= (PSTATE_BIT_UNDERWATER  * ((midPos.y +     radius) <  0.0f));
+	ps |= (PSTATE_BIT_UNDERWATER  * ((midPos.y +     radius) <  waterLevel));
 	ps |= (PSTATE_BIT_UNDERGROUND * ((midPos.y +     radius) <    gh));
 	ps |= (PSTATE_BIT_INAIR       * ((   pos.y -         wh) >   eps));
```

Comment:
`wh` probably needs to have the 0 replaced with water in its definiton.

Reply:
Any chance you know what `wh` actually stands for here? I'd love to rename these variable to something sane as well.

Reply:
It stands for the "ground height, no lower than the waterlevel."

Reply:
Can I call it `topSurfaceHeight`? It's the height of the upper most surface of the map that isn't air, and used solely for "is this in the air?" checks.